specifyweb/frontend/js_src/lib/components/WorkBench/WbUtils.tsx

```diff
@@ -132,101 +134,102 @@ export function WbUtilsComponent({
     utils.searchCells(
       event,
       searchRef.current as HTMLInputElement,
-      spreadsheetContainer.current
+      spreadsheetContainerRef?.current
     );
   };
 
   const handleReplace = (event: KeyboardEvent) => {
-    utils.replaceCells(event, replaceRef.current as HTMLInputElement);
+    utils.replaceCells(event, replaceRef.current!);
   };
 
   return (
-    <>
-      <div
-        aria-label={wbText.navigation()}
-        className="flex flex-wrap justify-end gap-x-1 gap-y-2"
-        role="toolbar"
-      >
-        <span className="contents" role="search">
+    <div
+      aria-label={wbText.navigation()}
+      className="flex flex-wrap justify-end gap-x-1 gap-y-2"
+      role="toolbar"
+    >
+      <span className="contents" role="search">
+        <div className="flex">
+          <Input.Generic
+            forwardRef={searchRef}
+            aria-label={commonText.searchQuery()}
+            autoComplete="on"
+            placeholder={commonText.search()}
+            spellCheck
+            title={commonText.searchQuery()}
+            type="search"
+            // @ts-expect-error KeyboardEvent type does not match?
+            onKeyDown={handleSearch}
+          />
+        </div>
+        {!isUploaded && hasPermission('/workbench/dataset', 'update') ? (
           <div className="flex">
-            <Input.Generic
-              forwardRef={searchRef}
-              aria-label={commonText.searchQuery()}
+            <Input.Text
```

Comment:
This should likely stay as Input.Generic because we want it to have `type="search"`

with type="search", a11y tech and browser shortcats would let users jump directly to the search field

Reply:
Thanks!

Also, getting this type error for the onKeyDown prop: `Type '(event: KeyboardEvent) => void' is not assignable to type 'KeyboardEventHandler<HTMLInputElement>'`

Any ideas?

If I change `event` to `KeyboardEventHandler<HTMLInputElement>` here and in `replaceCells`, then I get this instead when accessing `event.key`: `Property 'key' does not exist on type 'KeyboardEventHandler<HTMLInputElement>'`
https://github.com/specify/specify7/blob/a6790a826fc4c0b30276caeba784d8268bdccf51/specifyweb/frontend/js_src/lib/components/WorkBench/WbUtils.tsx#L141-L143

Reply:
The error I see in my IDE looks like this:
```yaml
Type '(event: KeyboardEvent) => void' is not assignable to type 'KeyboardEventHandler<HTMLInputElement>'.
  Types of parameters 'event' and 'event' are incompatible.
    Type 'KeyboardEvent<HTMLInputElement>' is not assignable to type 'KeyboardEvent'
```

the last line says it all - `Type 'KeyboardEvent<HTMLInputElement>' is not assignable to type 'KeyboardEvent'`

so you need to change this line:

```diff
- const handleReplace = (event: KeyboardEvent) => { 
+ const handleReplace = (event: KeyboardEvent<HTMLInputElement>) => {

Reply:
side note: some people prefer to define all callbacks and complex props before the `return ()` in in a component and then just pass those callbacks to the components as props.

the benefit of that is that the `return (...)` part (also known as template) stays cleaner, and just focused on rendering.

however, we generally don't do that - instead, most of the time we define callbacks inline inside the template.
pros:
- when reading the template, you don't have to constantly jump up and down to see the function definition and it's usage - it's definition and usage is on the same line.
- you don't have to do typescript typing manually - typescript will automatically infer typing for your function based on the function type that the prop expects

so in this case, it would look like this:

```diff
-  const handleReplace = (event: KeyboardEvent) => {
-    utils.replaceCells(event, replaceRef.current!);
-  };
- // ... later:
- onKeyDown={handleReplace}
+ onKeyDown={(event)=>utils.replaceCells(event, replaceRef.current!)}

this results in less code, less need to explicitly provide types, and less jumping around the file.

of course, if you need to memorize the function/value (useMemo/useCallback), or you want to reuse it in two callbacks, or you want to define it outside the component (to make testing easier or code more readable), or if the callback is very longer, then you are not forced to define the callback inline - but it does make things easier in a lot of cases.